Forgot Password

kakaly opened the pull request:
# Details

"Forgot Password" feature

## Description

* This PR adds "Forgot Password" page and completes the flow with
"Sent for recovery" page.
* Refactored "Login" page and moved out componenets to Utilities.js
for re-use.

## Fixes

This fixes https://github.com/Coding-Coach/coding-coach/issues/227

![forgotpassword_demo](https://user-images.githubusercontent.com/7032778/50379256-40490180-05fa-11e9-9cd7-fa3cd83386d7.gif)

kakaly commented:
@crysfel , @emmawedekind  - Let me know what you think. Not a great designer, so I could only do an average job for the svg component. I used an online convertor to convert the image to svg. Not sure how you would approach this.

tureey commented on src/pages/auth/ForgotPassword.js:
```diff
@@ -0,0 +1,115 @@
+            />
+            <div className="px-2 md:w-3/5 lg:w-1/3">
+              <Panel floating>
+                {this.state.displaySuccess ? this.sendResetLink() : this.getAccountInfo()}
```
Totally personal opinion: I prefer to have this encapsulated as a function
```
const renderDisplaySuccess = () =>
    this.state.displaySuccess
        ? this.sendResetLink()
        : this.getAccountInfo()
```

So we do:
`this.renderDisplaySuccess()`

kakaly replied:
Sure. I like this as well.

emmawedekind commented:
Love it! Can we leave space for the form feedback so the button doesn't have to shift down when the message is shown? :) Also it looks like the input box is cutting off the 'g' at the bottom :)

emmawedekind commented:
Also the "Oh no!" text under the header, can we increase the width? :)

emmawedekind commented on src/pages/auth/ForgotPassword.js:
```diff
@@ -0,0 +1,115 @@
+class ForgotPassword extends React.Component {
+  constructor(props) {
+    super(props);
+    this.getAccountInfo = this.getAccountInfo.bind(this);
```
Don't we still have to bind to the constructor? Or can you really use arrow functions for this?

Radi85 replied:
Arrow function will do this binding for us.

kakaly replied:
@tureey - Interesting point of view since React official documentation doesn't talk about this approach. I did a bit of reading about this and tried a few things locally. Here is my comparison between the two approaches

Arrow functions:
Pros:
1) No need to bind the function to the constructor

Cons:
When you use arrow functions, it does not get defined on the ProtoType. This has some consequences:
1) Child classes cannot call by, super.foo().
2) Overriding foo(), results in the creation of the function twice(one for super, one for derived)
3) This results in relatively slower performance compared to prototype approach.
4) It is hard to spy/stub/mock for testing while prototype functions can be easily mocked before instantiation.

I think the cons outweigh the pros here. I would stick with the current approach. But, I am interested in hearing more from you.

tureey replied:
About your points:
1 and 2. I never had to use them using React.
3. I don't know how much it really affects. Could you share some links talking about it, please? Because maybe I will it stop doing it.
4. I have never been a big fan of testing UI

tureey replied:
Well, even knowing this I would go for arrow functions, but just a total personal opinion.

Since you are using a callback to use `saveEmail` as `event => this.saveEmail(event)`. You're doing exactly the same than an arrow function, re assigning the function execution in run time.

So, you're having the same performance "issue". To evade this, you can `bind()` the function to the prototype as you commented but then to execute the function, you should just pass the reference to this as: ` onChange={this.saveEmail}`

kakaly replied:
Good catch. I modified it as per your suggestion. Thanks for pointing out. At this point, I am convinced about the arrow functions approach especially because it is cleaner. But, I would like others to pitch in and share their opinion too. If all of us agree with that approach, maybe we can open up a story to convert all components to use the arrow functions approach to have it uniform. :)

tureey replied:
Sure!! I would say though that it's a standard in the React community

emmawedekind commented on src/pages/auth/ForgotPassword.js:
```diff
@@ -0,0 +1,115 @@
+
+  clickHandler(e) {
+    e.preventDefault();
+    this.state.email === ''
```
Yes add .trim please!

kakaly replied:
Sure.

kakaly commented on src/pages/auth/ForgotPassword.js:
```diff
@@ -0,0 +1,115 @@
+  clickHandler(e) {
+    e.preventDefault();
+    this.state.email === ''
+      ? this.setState({ ...this.state, noinfo: true })
```
You're right. I do not remember why I wrote it that way. Doesn't make sense lol. Thanks for pointing out. I like the first approach better, as it is much more clear as to what we are doing. The second one is very declarative, which is great. Just a personal preference for the first one.

tureey replied:
Your call!

kakaly commented on src/pages/auth/ForgotPassword.js:
```diff
@@ -0,0 +1,115 @@
+  }
+
+  saveEmail(e) {
+    this.setState({ ...this.state, email: e.target.value });
```
You got it!

kakaly commented on src/pages/auth/ForgotPassword.js:
```diff
@@ -0,0 +1,115 @@
+              <TextField
+                label={i18n._(t('auth.signup.email')`Email Address`)}
+                name="email"
+                onChange={(event) => this.saveEmail(event)}
```
Sure. This is probably another advantage of using arrow function. But, like I mentioned previously, I am not quite sure if its worth losing key things like mocking/stubbing for a step improvement in cleanliness. I maybe wrong too. Let me know what you think.

kakaly commented:
@emmawedekind - I have addressed your comments. Like I said, I am not a great designer. So, comments like these really help. Let me know if you think there is anything else that needs attention.

@tureey - Thanks for looking at my code. Learnt a bunch from your perspective.

tureey commented:
We are here to learn! I think you did a great job.

Well done!😊

kakaly commented:
@emmawedekind - I would like to take on something new for the front end. Feel free to assign me anything cos I couldn't figure out any open issues from the issues tab.

crysfel commented on src/pages/auth/Login.js:
```diff
@@ -3,43 +3,32 @@ import { Trans, t } from '@lingui/macro';
                     <Trans id="auth.signin.title" defaults="Sign In" />
                   </h2>
-                  <a href="#noop" className="text-sm h-4 mt-1">
+                  <a href="recovery" className="text-sm h-4 mt-1">
```
Please use the `Link` component from `reach-router`.

tureey replied:
Stupid question but, which is the benefit of it? In my side projects I use the `<a>` too! Could you light up a little bit please @crysfel ?

crysfel replied:
If using an `a` element, there will be a full reload of the next page (Which is not ideal). If using `Link` it will only render the parts that are changing.

tureey replied:
basically the `<a>` will execute a window.location or similar... but I thought that `Link` would do the same... so does it keep references from nodes and things like this? @crysfel

crysfel replied:
You would use `<a>` when linking to a different website. And `Link` for internal links, because it doesn't reload the react app, it works with the `History` api: https://css-tricks.com/using-the-html5-history-api/#article-header-id-1

tureey replied:
Tomorrow I'll check this. Thank you so much @crysfel  :smile:

kakaly replied:
Good information @crysfel . I believe since Link is part of React, it has the intelligence to re render only the parts that changed. Good to know.

crysfel commented on src/pages/auth/ForgotPassword.js:
```diff
@@ -0,0 +1,118 @@
+import TextField from 'components/textfield/TextField';
+import { Panel, PanelContent } from 'components/panel/Panel';
+import { SignUp, TitleMessage } from './Utilities';
+import { ReactComponent as HandImage } from '../home/assets/images/forgot-password.svg';
```
Please move `forgot-password.svg` to `./assets/images/forgot-password.svg`. Each module should have their own images.

I'm also thinking on creating a common assets folder for the logo and other common images. But the idea is that each module should have their own unique assets (if any)

tureey replied:
I agree that we need a common

kakaly commented:
@crysfel , @tureey - Refactored the component to use arrow functions now :) and @crysfel I have taken care of your other comments too. Please take a look. Thanks.

crysfel commented on src/pages/auth/ForgotPassword.js:
```diff
@@ -0,0 +1,115 @@
+class ForgotPassword extends React.Component {
+  constructor(props) {
+    super(props);
+    this.state = { displaySuccess: false, email: '', noinfo: false };
```
Just for code consistency across the project, remove the `constructor` and define the state like this:

```
class ForgotPassword extends React.Component {
  state = {
    displaySuccess: false,
    email: '',
    noinfo: false,
  };

  //.....
```

kakaly replied:
Sure. Updated!
